core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java

```diff
@@ -526,7 +526,7 @@ private void cleanFailedShards(final ClusterChangedEvent event) {
                         if (localNodeId.equals(shardRouting.currentNodeId())) {
                             // we have a timeout here just to make sure we don't have dangled failed shards for some reason
                             // its just another safely layer
-                            if (shardRouting.version() == failedShard.version && ((now - failedShard.timestamp) < TimeValue.timeValueMinutes(60).millis())) {
+                            if (shardRouting.allocationId().getId().equals(failedShard.allocationId) && ((now - failedShard.timestamp) < TimeValue.timeValueMinutes(60).millis())) {
```

Comment:
I don't think we need this FailedShard class anymore - that time check is superfluous - we can just use Set<String> with allocation ids. Can you put it on your hit list and remove it in another PR?